### PR TITLE
Let Ads know the status of their re-size request via  embed-size-denied and embed-size-changed

### LINKED ISSUE #1533: Implement embed-size-denied
signal back ads/iframes when we couldn't expand on their embed-resize request. Send a postMessage with embed-size-denied (when the runtime can't resize)so that they can provide their own overflow element within and repeat request on tap. In that case AMP will definitely allow size change due to user action.

This would be useful for both: iframes and ads

CC @dvoytenko, @cramforce

### PR DESCRIPTION
**Pull request by sriramkrish85**
Fixes #1533

### CODE REVIEW

**Comment by ampsauce** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 8053a8aad905ba62765d5dec969e9d3bb9c9f1a5
Build details: https://travis-ci.org/ampsauce/amphtml/builds/105007633

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: 3p/integration.js [hunk 1]
```diff
@@ -126,6 +126,7 @@ window.draw3p = function(opt_configCallback) {
 
   // This only actually works for ads.
   window.context.observeIntersection = observeIntersection;
+  window.context.observeResize = observeResize;
```
Comment (cramforce) [inline 3p/integration.js]: Been thinking for a while and I think it is better to just make this two functions
`onResizeSuccess` and `onResizeDenied`

File: builtins/amp-ad.md [hunk 2]
```diff
@@ -67,19 +67,11 @@ Optional attribute to pass configuration to the ad as an arbitrarily complex JSO
 An `amp-ad` must have static layout defined as is the case with any other AMP element. However,
 it's possible to resize an `amp-ad` in runtime. To do so:
 
-1. The `amp-ad` must be defined with `resizable` attribute;
-2. The `amp-ad` must have `overflow` child element;
-3. The Ad's Iframe document has to send a `embed-size` request as a window message.
+1. The Ad's Iframe document has to send a `embed-size` request as a window message.
```
Comment (cramforce) [inline builtins/amp-ad.md]: I think we can just delete this here. No need to document how this works to users.

**Comment by cramforce** [comment]
Just some comments.

File: 3p/integration.js (hunk 1 above)
Comment (sriramkrish85) [inline 3p/integration.js]: Done

File: builtins/amp-ad.md (hunk 2 above)
Comment (sriramkrish85) [inline builtins/amp-ad.md]: Done

**Comment by sriramkrish85** [comment]
PTAL

File: ads/README.md [hunk 3]
```diff
@@ -91,9 +91,31 @@ window.parent.postMessage({
...
 possible, but it will take into account where the reader is currently reading, whether the scrolling
-is ongoing and any other UX or performance factors. If the runtime cannot satisfy the resize events
-the `amp-ad` will show an `overflow` element. Clicking on the `overflow` element will immediately
-resize the `amp-ad` since it's triggered by a user action.
+is ongoing and any other UX or performance factors.
+
+Based one whether the AMP runtime was able to satisfy the resize request,
+the runtime will also send out a `embed-size-changed` or `embed-size-denied` message accordingly. The ad can listen to these messages and have it's own overflow element within and repeat request on tap. In that case AMP runtime will definitely allow size change due to user action.
+
+Ads can call the special API `window.context.onResizeSuccess` to see if a resize request was successful.
```
Comment (cramforce) [inline ads/README.md]: Here and below instead of "to see if", "to get a callback in case…"

**Comment by cramforce** [comment]
LGTM

File: builtins/amp-ad.js [hunk 4]
```diff
@@ -288,19 +276,34 @@ export function installAd(win) {
       }
     }
 
+    /** @override  */
+    onOverflowCallback(overflown, requestedHeight) {
+      const targetOrigin =
+          this.iframe_.src ? parseUrl(this.iframe_.src).origin : '*';
+      postMessage(
```
Comment (dvoytenko) [inline builtins/amp-ad.js]: This should be done only `if (overflown)`.

File: src/base-element.js [hunk 5]
```diff
@@ -587,4 +587,11 @@ export class BaseElement {
    * @protected
    */
   onLayoutMeasure() {}
+
+  /**
+   * Called after a overflowCallback is triggered on an element.
+   * @param {boolean} unusedOverflown
+   * @param {number} unusedRequestedHeight
+   */
+  onOverflowCallback(unusedOverflown, unusedRequestedHeight) {}
```
Comment (dvoytenko) [inline src/base-element.js]: Just `overflowCallback`

File: src/custom-element.js [hunk 6]
```diff
@@ -1033,38 +1033,32 @@ export function createAmpElementProto(win, name, implementationClass) {
...
    */
   ElementProto.overflowCallback = function(overflown, requestedHeight) {
-    if (!overflown && !this.overflowElement_) {
-      // Overflow has never been initialized and not wanted.
-      return;
-    }
-
-    const overflowElement = this.getOverflowElement();
-    if (!overflowElement) {
+    this.getOverflowElement();
```
Comment (dvoytenko) [inline src/custom-element.js]: Generally, you can keep all the code intact if you just put the `implementation.overflowCallback` call up top.

File: builtins/amp-ad.js [hunk 7]
```diff
@@ -22,7 +22,6 @@ import {IntersectionObserver} from '../src/intersection-observer';
 import {isLayoutSizeDefined} from '../src/layout';
```
Comment (dvoytenko) [inline builtins/amp-ad.js]: Are you also planning to do the same for `amp-iframe`?

File: builtins/amp-ad.js (hunk 4 above)
Comment (sriramkrish85) [inline builtins/amp-ad.js]: Done

File: ads/README.md (hunk 3 above)
Comment (sriramkrish85) [inline ads/README.md]: Done

File: src/custom-element.js (hunk 6 above)
Comment (sriramkrish85) [inline src/custom-element.js]: I'd prefer if the generic element code to run first before the custom overridden code executes.

File: src/base-element.js (hunk 5 above)
Comment (sriramkrish85) [inline src/base-element.js]: Done

File: builtins/amp-ad.js (hunk 7 above)
Comment (sriramkrish85) [inline builtins/amp-ad.js]: per @cramforce, we would still like to have the resizable attribute and overflow element. So not sure if these would still be helpful for `amp-iframe`.

**Comment by sriramkrish85** [comment]
PTAL

File: builtins/amp-ad.js (hunk 7 above)
Comment (dvoytenko) [inline builtins/amp-ad.js]: I believe this would still be rather helpful there as well.

Comment (dvoytenko) [inline builtins/amp-ad.js]: So, let's create a bug to track it. It doesn't have to be in this PR.

File: test/functional/test-custom-element.js [hunk 8]
```diff
@@ -1125,19 +1125,27 @@ describe('CustomElement Overflow Element', () => {
   });
 
   it('should set overflow', () => {
+    const overflowCallbackSpy =
+        sinon.spy(element.implementation_, 'overflowCallback');
```
Comment (dvoytenko) [inline test/functional/test-custom-element.js]: Use `sandbox.spy`. Then you don't need `restore` below.

**Comment by dvoytenko** [comment]
LGTM with a minor comment.

File: test/functional/test-custom-element.js (hunk 8 above)
Comment (sriramkrish85) [inline test/functional/test-custom-element.js]: Done

**Pull request merged**